### PR TITLE
fix(libcontainer) no_pivot args is not used

### PR DESCRIPTION
**Pull request by xujihui1985**
fix the problem that no_pivot args is not used when create container when we prepare rootfs with chroot, we should move_mount the rootfs before chroot, otherwise we will not able to use the new rootfs when exec into the container

### CODE REVIEW

**Comment by xujihui1985** [comment]
@udzura This PR addresses the issue where the no_pivot argument is not recognized. This argument is crucial in environments where the pivot_root syscall is not permitted, such as when running containers within containers.

**Comment by udzura** [comment]
I understand what this PR aim to do 👍  But I have no privilege to merge this. I guess you've mistaken me for someone!

**Comment by xujihui1985** [comment]
> I understand what this PR aim to do 👍  But I have no privilege to merge this. I guess you've mistaken me for someone!

@udzura sorry for bother you, could you help to involve someone that is able to have a look at this PR? Really appreciate.

**Comment by utam0k** [comment]
I'll check this PR. Thanks for your first contribution!

**Comment by utam0k** [comment]
@xujihui1985 May I ask you to add an integration test for this PR?
https://containers.github.io/youki/developer/e2e/rust_oci_test.html

**Comment by xujihui1985** [comment]
> @xujihui1985 May I ask you to add an integration test for this PR?
> 
> https://containers.github.io/youki/developer/e2e/rust_oci_test.html

sure, I'd like to learn how to add integrate test

**Comment by utam0k** [comment]
> sure, I'd like to learn how to add integrate test

I can help you if you need ;)

**Comment by YJDoc2** [comment]
Hey @xujihui1985 Thanks for your contribution! May I ask you to also take a look at https://github.com/containers/youki/pull/2597 , which also deals with `no_pivot` and the create and run commands. Can you check if you need to pick up any changes from there to here? Thanks.

**Comment by xujihui1985** [comment]
@YJDoc2 I have pick the change from commit https://github.com/containers/youki/pull/2597, I will continue work on integration test

**Comment by xujihui1985** [comment]
@udzura Hi, I have implemented the integration test in commit [c82496b](https://github.com/containers/youki/pull/2923/commits/c82496b64539d0b776a82022cca1979e448e5b9e). For now, I copied and reused some utility functions, like test_inside_container -> test_inside_container_with_no_pivot, since create_container currently doesn't allow setting the necessary arguments. and change this function will cause lots of change in current testcase. This was done to verify if the test works as expected. Once the test is reviewed and approved, I'll refactor the functions accordingly.

File: tests/contest/contest/src/tests/no_pivot/mod.rs
Comment (utam0k) [inline tests/contest/contest/src/tests/no_pivot/mod.rs:1]: That's what I expected. Please go ahead to refactoring 🙏

Comment (xujihui1985) [reply tests/contest/contest/src/tests/no_pivot/mod.rs:1]: sure, let me know if any improvements are possible

File: crates/libcontainer/src/process/container_init_process.rs
Comment (YJDoc2) [inline crates/libcontainer/src/process/container_init_process.rs:273]: Can  you add a doc comment for this function, explaining the intention behind it and use?

File: tests/contest/runtimetest/src/tests.rs
Comment (YJDoc2) [inline tests/contest/runtimetest/src/tests.rs:574]: Also add a doc comment explaining the intent behind this test please.

**Review by YJDoc2 — COMMENTED** [review]
Hey, the overall implementation looks good, I have a couple of nitpick comments, please take a look at them.

As for the tests, I feel that the added test is ok, but we need to refactor the `test_inside...` function to avoid duplication. I'll think on a way for more "ergonomic" imple, but if you can think of something, go ahead.

I'm also thinking we should have this functionality in `test_outside..` so we can also test for container create,run,delete with no_pivot. WDYT? Would that be useful?

File: crates/libcontainer/src/process/container_init_process.rs
Comment (xujihui1985) [reply crates/libcontainer/src/process/container_init_process.rs:301]: ya, the logic is borrowed from crun and similar logic is implement in runc as well https://github.com/containers/crun/blob/53cd1c1c697d7351d0cad23708d29bf4a7980a3a/src/libcrun/linux.c#L2780
the reason behind these is to umount or hide the proc and sys fs before change it's root to new rootfs, umount here is safe because procfs and sysfs is privately mounted in the containers mountns, this is not required when change rootfs with pivotroot as the old rootfs will be swap to different location

**Comment by xujihui1985** [comment]
> Hey, the overall implementation looks good, I have a couple of nitpick comments, please take a look at them.
> 
> As for the tests, I feel that the added test is ok, but we need to refactor the `test_inside...` function to avoid duplication. I'll think on a way for more "ergonomic" imple, but if you can think of something, go ahead.
> 
> I'm also thinking we should have this functionality in `test_outside..` so we can also test for container create,run,delete with no_pivot. WDYT? Would that be useful?

@YJDoc2  I believe the test_outside_container with the no-pivot option doesn't behave any differently from the test with the option enabled. It's unclear what to assert when testing outside the container with the no-pivot option because, outside of a container mount namespace, chroot and pivot_root have no noticeable effect. Any idea?

**Comment by xujihui1985** [comment]
@YJDoc2 Do you think it would be better to refactor the test_inside... function in a separate PR? It will involve a lot of changes as the change of the signature

**Comment by xujihui1985** [comment]
> @YJDoc2 Do you think it would be better to refactor the test_inside... function in a separate PR? It will involve a lot of changes as the change of the signature

@YJDoc2 Hi, would you like to proceed the PR or wait until the function been refactored?

**Comment by YJDoc2** [comment]
> I believe the test_outside_container with the no-pivot option doesn't behave any differently from the test with the option enabled. It's unclear what to assert when testing outside the container with the no-pivot option because, outside of a container mount namespace, chroot and pivot_root have no noticeable effect. 

I mentioned `test_outside_container` as that is another test fn we use, but I hadn't thought this in detail. My main intention here was that to verify via e2e that basic commands such as start, delete, run etc work properly with the no_pivot, so whatever does the trick, works.

> @YJDoc2 Hi, would you like to proceed the PR or wait until the function been refactored?

I think the refactoring should be done in separate PR as that will be a big change, I'll do a final review, hopefully today, and if all's ok then merge, after which you can start with that refactoring.

**Comment by xujihui1985** [comment]
> > I believe the test_outside_container with the no-pivot option doesn't behave any differently from the test with the option enabled. It's unclear what to assert when testing outside the container with the no-pivot option because, outside of a container mount namespace, chroot and pivot_root have no noticeable effect. 
> 
> 
> 
> I mentioned `test_outside_container` as that is another test fn we use, but I hadn't thought this in detail. My main intention here was that to verify via e2e that basic commands such as start, delete, run etc work properly with the no_pivot, so whatever does the trick, works.
> 
> 
> 
> > @YJDoc2 Hi, would you like to proceed the PR or wait until the function been refactored?
> 
> 
> 
> I think the refactoring should be done in separate PR as that will be a big change, I'll do a final review, hopefully today, and if all's ok then merge, after which you can start with that refactoring.

@YJDoc2 Hi, hope everything goes fine with you, if there is anything I can improve for this PR, please let me know

File: crates/libcontainer/src/process/container_init_process.rs
Comment (YJDoc2) [inline crates/libcontainer/src/process/container_init_process.rs:351]: nit: we should move the comment above this to the function where we are doing this

**Review by YJDoc2 — APPROVED** [review]
Hey, I have a couple of nits, but nothing that blocks the merge. I'm approving this PR, and you can address the nits in the PR you'll make to refactor the test function. I'm not merging this right away, in case you want to reply something, but if no reply, I'll merge it by tomorrow.

**Comment by YJDoc2** [comment]
> @YJDoc2 Hi, hope everything goes well with you, if there is anything I can improve for this PR, please let me know as

Hey, sorry, I was busy with work and some other stuff and couldn't get to any of this. This is good to merge, and I have approved this. I will wait till tom if you have any replies to my comments, otherwise I'll merge this as-is, and you can address the nits in your next PR.

**Comment by xujihui1985** [comment]
> > @YJDoc2 Hi, hope everything goes well with you, if there is anything I can improve for this PR, please let me know as
> 
> 
> 
> Hey, sorry, I was busy with work and some other stuff and couldn't get to any of this. This is good to merge, and I have approved this. I will wait till tom if you have any replies to my comments, otherwise I'll merge this as-is, and you can address the nits in your next PR.

thanks for your reply. I will continue working on the refactoring. Cheers

**Comment by YJDoc2** [comment]
>  I will continue working on the refactoring. 

Thank you! Please address the nitpicks in your next PR.

**Pull request merged**
